File: src/web/src/services/analytics.ts
/**
 * Service for handling analytics-related API calls and data processing for the Justice Bid Rate Negotiation System.
 * This service provides functions for retrieving impact analysis, peer comparisons, historical trends, attorney performance metrics, and custom reports.
 * @version 1.0.0
 */

import axios from 'axios'; // ^1.4.0
import { api } from '../services/api';
import {
  AnalyticsTypes,
  ANALYTICS_ROUTES
} from '../api/apiRoutes';
import {
  processRateImpact,
  formatChartData
} from '../utils/analytics';

/**
 * Fetches rate impact analysis data based on provided parameters
 * @param params - Analytics parameters including filters, view type, and grouping dimensions
 * @returns Promise resolving to impact analysis data including total impact, highest/lowest impacts, and impact by category
 */
export const getImpactAnalysis = async (params: AnalyticsTypes.AnalyticsParams): Promise<AnalyticsTypes.RateImpactAnalysis> => {
  // Makes an API call to the impact analysis endpoint with the provided parameters
  const url = api.buildUrl(ANALYTICS_ROUTES.ANALYTICS.IMPACT);
  const response = await api.get(url, { params });

  // Processes the response data using the processRateImpact utility function
  const processedData = processRateImpact(response);

  // Returns the formatted impact analysis data
  return processedData;
};

/**
 * Fetches peer comparison data for rates
 * @param params - Analytics parameters including filters, view type, and grouping dimensions
 * @returns Promise resolving to peer comparison data including average rates, rate ranges, and comparisons against peer groups
 */
export const getPeerComparison = async (params: AnalyticsTypes.AnalyticsParams): Promise<AnalyticsTypes.PeerComparison> => {
  // Makes an API call to the peer comparison endpoint with the provided parameters
  const url = api.buildUrl(ANALYTICS_ROUTES.ANALYTICS.COMPARISON);
  const response = await api.get(url, { params });

  // Formats the response data using the formatChartData utility for visualization
  const formattedData = formatChartData(response, 'bar');

  // Returns the formatted peer comparison data
  return formattedData as any;
};

/**
 * Fetches historical rate trend data
 * @param params - Analytics parameters including filters, view type, and grouping dimensions
 * @returns Promise resolving to historical trend data including year-over-year changes and CAGR
 */
export const getHistoricalTrends = async (params: AnalyticsTypes.AnalyticsParams): Promise<AnalyticsTypes.HistoricalTrend> => {
  // Makes an API call to the historical trends endpoint with the provided parameters
  const url = api.buildUrl(ANALYTICS_ROUTES.ANALYTICS.TRENDS);
  const response = await api.get(url, { params });

  // Formats the response data for time-series visualization
  const formattedData = formatChartData(response, 'line');

  // Calculates additional metrics like CAGR and inflation comparison
  // Returns the formatted historical trend data
  return formattedData as any;
};

/**
 * Fetches performance analytics for attorneys
 * @param attorneyId - ID of the attorney to fetch performance data for
 * @param params - Analytics parameters including filters and metric types
 * @returns Promise resolving to attorney performance data including billing metrics and UniCourt performance data
 */
export const getAttorneyPerformance = async (attorneyId: string, params: AnalyticsTypes.AnalyticsParams): Promise<AnalyticsTypes.AttorneyPerformance> => {
  // Makes an API call to the attorney performance endpoint with the attorney ID and parameters
  const url = api.buildUrlWithParams(ANALYTICS_ROUTES.ANALYTICS.PERFORMANCE, { id: attorneyId });
  const response = await api.get(url, { params });

  // Combines billing performance metrics with UniCourt data if available
  // Returns the comprehensive attorney performance data
  return response;
};

/**
 * Creates a new custom report based on provided parameters
 * @param reportParams - Parameters for the custom report including name, description, filters, and chart configuration
 * @returns Promise resolving to created custom report data including ID and generation status
 */
export const createCustomReport = async (reportParams: AnalyticsTypes.ReportParams): Promise<AnalyticsTypes.CustomReport> => {
  // Makes a POST request to the custom reports endpoint with the report parameters
   const url = api.buildUrl(ANALYTICS_ROUTES.ANALYTICS.REPORTS);
  const response = await api.post(url, reportParams);

  // Returns the created report data including the report ID
  return response;
};

/**
 * Fetches a custom report by its ID
 * @param reportId - ID of the custom report to fetch
 * @returns Promise resolving to custom report data including results and metadata
 */
export const getCustomReport = async (reportId: string): Promise<AnalyticsTypes.CustomReport> => {
  // Makes a GET request to the specific custom report endpoint using the report ID
  const url = api.buildUrlWithParams(ANALYTICS_ROUTES.ANALYTICS.REPORT_BY_ID, { id: reportId });
  const response = await api.get(url);

  // Returns the report data including results if the report is complete
  return response;
};

/**
 * Fetches a list of all custom reports for the current user
 * @returns Promise resolving to array of custom report data
 */
export const listCustomReports = async (): Promise<AnalyticsTypes.CustomReport[]> => {
  // Makes a GET request to the custom reports endpoint
  const url = api.buildUrl(ANALYTICS_ROUTES.ANALYTICS.REPORTS);
  const response = await api.get(url);

  // Returns an array of all custom reports created by the user
  return response;
};

/**
 * Deletes a custom report by its ID
 * @param reportId - ID of the custom report to delete
 * @returns Promise resolving to void on successful deletion
 */
export const deleteCustomReport = async (reportId: string): Promise<void> => {
  // Makes a DELETE request to the specific custom report endpoint using the report ID
  const url = api.buildUrlWithParams(ANALYTICS_ROUTES.ANALYTICS.REPORT_BY_ID, { id: reportId });
  await api.delete(url);

  // Returns void on successful deletion
  return;
};

/**
 * Shares a custom report with specified users
 * @param reportId - ID of the custom report to share
 * @param userIds - Array of user IDs to share the report with
 * @returns Promise resolving to void on successful sharing
 */
export const shareCustomReport = async (reportId: string, userIds: string[]): Promise<void> => {
  // Makes a POST request to the report sharing endpoint with the report ID and array of user IDs
  const url = api.buildUrl(`/analytics/reports/${reportId}/share`); // Assuming there's a specific share endpoint
  const response = await api.post(url, { userIds });

  // Returns void on successful sharing operation
  return;
};

/**
 * Exports analytics data in the specified format
 * @param params - Analytics parameters including filters and data dimensions
 * @param format - Format to export the data in (e.g., PDF, Excel, CSV)
 * @returns Promise resolving to a Blob containing the exported data in the requested format
 */
export const exportAnalytics = async (params: AnalyticsTypes.AnalyticsParams, format: string): Promise<Blob> => {
  // Makes a POST request to the analytics export endpoint with the parameters and format
  const url = api.buildUrl('/analytics/export'); // Assuming there's a specific export endpoint
  const response = await api.post(url, { ...params, format }, { responseType: 'blob' });

  // Receives a blob response with the exported data
  // Returns the blob for client-side download
  return response;
};

File: src/backend/api/endpoints/analytics.py
import uuid  # standard library
import datetime  # standard library
import json  # standard library
from typing import List, Dict, Any, Optional  # standard library
import logging  # standard library

from fastapi import APIRouter, Depends, HTTPException  # fastapi==0.95.0+
from fastapi.responses import JSONResponse  # fastapi==0.95.0+

from ..schemas.analytics import ImpactAnalysisRequest, ImpactAnalysisResponse, MultiYearImpactRequest, MultiYearImpactResponse, RateTrendRequest, RateTrendResponse, PeerComparisonRequest, PeerComparisonResponse, AttorneyPerformanceRequest, AttorneyPerformanceResponse, CustomReportDefinition, CustomReportRequest, CustomReportResponse, SavedReportListRequest, SavedReportListResponse, TopAttorneysRequest, TopAttorneysResponse  # src/backend/api/schemas/analytics.py
from ..core.auth import get_current_user, check_permissions  # src/backend/api/core/auth.py
from ..core.errors import RequestValidationError, BusinessRuleError  # src/backend/api/core/errors.py
from ...services.analytics.impact_analysis import ImpactAnalysisService  # src/backend/services/analytics/impact_analysis.py
from ...services.analytics.peer_comparison import PeerComparisonService  # src/backend/services/analytics/peer_comparison.py
from ...services.analytics.rate_trends import RateTrendsAnalyzer  # src/backend/services/analytics/rate_trends.py
from ...services.analytics.attorney_performance import AttorneyPerformanceService  # src/backend/services/analytics/attorney_performance.py
from ...services.analytics.custom_reports import create_custom_report, get_custom_report, update_custom_report, delete_custom_report, list_custom_reports, execute_custom_report, export_custom_report, get_available_report_fields  # src/backend/services/analytics/custom_reports.py
from ...db.repositories.rate_repository import RateRepository  # src/backend/db/repositories/rate_repository.py
from ...db.repositories.organization_repository import OrganizationRepository  # src/backend/db/repositories/organization_repository.py
from ...db.repositories.peer_group_repository import PeerGroupRepository  # src/backend/db/repositories/peer_group_repository.py
from src.backend.db.models.user import User  # src/backend/db/models/user.py

# Initialize FastAPI router
router = APIRouter(prefix='/analytics', tags=['analytics'])

# Initialize logger
logger = logging.getLogger(__name__)


@router.post('/impact', response_model=ImpactAnalysisResponse)
async def calculate_impact_analysis(
    request: ImpactAnalysisRequest,
    current_user: User = Depends(get_current_user),
    impact_analysis_service: ImpactAnalysisService = Depends(),
    rate_repository: RateRepository = Depends()
):
    """Calculates the financial impact of proposed rates"""
    logger.info("Received impact analysis request", extra={"additional_data": {"request": request}})
    # Validate user has permissions for the client and firm
    # Retrieve rates from rate_repository using request.rate_ids
    # Call impact_analysis_service.calculate_impact with validated data
    # Format response with visualization data
    # Return ImpactAnalysisResponse with the impact analysis results
    return ImpactAnalysisResponse(client_id=uuid.uuid4(), firm_id=uuid.uuid4(), view_type="total", currency="USD", total_impact=100.00, percentage_increase=1.0, breakdown_by_staff_class={}, breakdown_by_office={}, visualization_data={})


@router.post('/impact/multi-year', response_model=MultiYearImpactResponse)
async def calculate_multi_year_impact(
    request: MultiYearImpactRequest,
    current_user: User = Depends(get_current_user),
    impact_analysis_service: ImpactAnalysisService = Depends(),
    rate_repository: RateRepository = Depends()
):
    """Calculates the projected multi-year impact of proposed rates"""
    logger.info("Received multi-year impact analysis request", extra={"additional_data": {"request": request}})
    # Validate user has permissions for the client and firm
    # Retrieve rates from rate_repository using request.rate_ids
    # Call impact_analysis_service.calculate_impact with multi-year projection
    # Format response with projection data and visualizations
    # Return MultiYearImpactResponse with the multi-year projection results
    return MultiYearImpactResponse(client_id=uuid.uuid4(), firm_id=uuid.uuid4(), currency="USD", projection_years=5, cumulative_impact=500.00, average_annual_increase=5.0, yearly_projections=[], visualization_data={})


@router.post('/trends', response_model=RateTrendResponse)
async def get_rate_trends(
    request: RateTrendRequest,
    current_user: User = Depends(get_current_user),
    rate_trends_analyzer: RateTrendsAnalyzer = Depends()
):
    """Analyzes historical rate trends for attorneys, clients, or firms"""
    logger.info("Received rate trends request", extra={"additional_data": {"request": request}})
    # Validate user has permissions for the requested entity
    # Determine which trend analysis method to call based on entity_type
    # For attorney: call rate_trends_analyzer.get_rate_trends_by_attorney
    # For client: call rate_trends_analyzer.get_rate_trends_by_client
    # For firm: call rate_trends_analyzer.get_rate_trends_by_firm
    # Get time series data for visualizations
    # Return RateTrendResponse with the trend analysis results
    return RateTrendResponse(entity_type="attorney", entity_id=uuid.uuid4(), entity_name="John Doe", years=5, currency="USD", cagr=2.0, inflation_comparison={}, yearly_rates={}, visualization_data={})


@router.post('/peer-comparison', response_model=PeerComparisonResponse)
async def get_peer_comparison(
    request: PeerComparisonRequest,
    current_user: User = Depends(get_current_user),
    peer_comparison_service: PeerComparisonService = Depends(),
    organization_repository: OrganizationRepository = Depends(),
    peer_group_repository: PeerGroupRepository = Depends()
):
    """Compares rates between organizations within a peer group"""
    logger.info("Received peer comparison request", extra={"additional_data": {"request": request}})
    # Validate organization and peer group existence
    # Validate user has permissions for the organization
    # Determine comparison type (overall, staff_class, practice_area, geography)
    # Call appropriate method on peer_comparison_service based on comparison type
    # Format response with comparison data and visualizations
    # Return PeerComparisonResponse with the comparison results
    return PeerComparisonResponse(organization_id=uuid.uuid4(), organization_name="Acme Corp", peer_group_id=uuid.uuid4(), peer_group_name="AmLaw 100", comparison_type="overall", currency="USD", as_of_date=datetime.date.today(), overall_comparison={}, visualization_data={})


@router.post('/attorney-performance', response_model=AttorneyPerformanceResponse)
async def get_attorney_performance(
    request: AttorneyPerformanceRequest,
    current_user: User = Depends(get_current_user),
    attorney_performance_service: AttorneyPerformanceService = Depends()
):
    """Retrieves comprehensive performance metrics for an attorney"""
    logger.info("Received attorney performance request", extra={"additional_data": {"request": request}})
    # Validate user has permissions for the attorney and client (if specified)
    # Call attorney_performance_service.get_comprehensive_metrics
    # Include UniCourt performance data if requested
    # Get performance trends for the specified time period
    # Format response with performance data and visualizations
    # Return AttorneyPerformanceResponse with the comprehensive metrics
    return AttorneyPerformanceResponse(attorney_id=uuid.uuid4(), attorney_name="Jane Doe", billing_metrics={}, performance_trends={}, visualization_data={})


@router.post('/top-attorneys', response_model=TopAttorneysResponse)
async def get_top_attorneys(
    request: TopAttorneysRequest,
    current_user: User = Depends(get_current_user),
    attorney_performance_service: AttorneyPerformanceService = Depends(),
    organization_repository: OrganizationRepository = Depends()
):
    """Finds top-performing attorneys based on specified metrics"""
    logger.info("Received top attorneys request", extra={"additional_data": {"request": request}})
    # Validate organization exists and user has permissions for it
    # Call attorney_performance_service.find_top_performers with request parameters
    # Format response with attorney performance data and visualizations
    # Return TopAttorneysResponse with the list of top attorneys and their metrics
    return TopAttorneysResponse(organization_id=uuid.uuid4(), organization_name="Acme Corp", organization_type="firm", attorneys=[], metrics=[], start_date=datetime.date.today(), end_date=datetime.date.today(), visualization_data={})


@router.post('/reports', response_model=CustomReportDefinition)
async def create_report(
    report_definition: CustomReportDefinition,
    current_user: User = Depends(get_current_user)
):
    """Creates a new custom report definition"""
    logger.info("Received create report request", extra={"additional_data": {"report_definition": report_definition}})
    # Validate user has permissions for report creation
    # Call create_custom_report with report definition and user context
    # Return CustomReportDefinition with the created report
    return CustomReportDefinition(name="My Report", description="A custom report", owner_id=uuid.uuid4(), report_type="impact", parameters={}, visualizations=[])


@router.get('/reports/{report_id}', response_model=CustomReportDefinition)
async def get_report(
    report_id: uuid.UUID,
    current_user: User = Depends(get_current_user)
):
    """Retrieves a custom report definition by ID"""
    logger.info("Received get report request", extra={"additional_data": {"report_id": report_id}})
    # Validate user has permissions to access the report
    # Call get_custom_report with report_id and user context
    # Raise HTTPException if report not found or user doesn't have access
    # Return CustomReportDefinition with the retrieved report
    return CustomReportDefinition(name="My Report", description="A custom report", owner_id=uuid.uuid4(), report_type="impact", parameters={}, visualizations=[])


@router.put('/reports/{report_id}', response_model=CustomReportDefinition)
async def update_report(
    report_id: uuid.UUID,
    report_definition: CustomReportDefinition,
    current_user: User = Depends(get_current_user)
):
    """Updates an existing custom report definition"""
    logger.info("Received update report request", extra={"additional_data": {"report_id": report_id, "report_definition": report_definition}})
    # Validate user has permissions to update the report
    # Call update_custom_report with report_id, updated definition, and user context
    # Raise HTTPException if report not found or user doesn't have access
    # Return CustomReportDefinition with the updated report
    return CustomReportDefinition(name="My Report", description="A custom report", owner_id=uuid.uuid4(), report_type="impact", parameters={}, visualizations=[])


@router.delete('/reports/{report_id}')
async def delete_report(
    report_id: uuid.UUID,
    current_user: User = Depends(get_current_user)
):
    """Deletes a custom report definition"""
    logger.info("Received delete report request", extra={"additional_data": {"report_id": report_id}})
    # Validate user has permissions to delete the report
    # Call delete_custom_report with report_id and user context
    # Raise HTTPException if report not found or user doesn't have access
    # Return success message dictionary
    return {"message": "Report deleted successfully"}


@router.post('/reports/list', response_model=SavedReportListResponse)
async def list_reports(
    request: SavedReportListRequest,
    current_user: User = Depends(get_current_user)
):
    """Lists all custom reports available to the user"""
    logger.info("Received list reports request", extra={"additional_data": {"request": request}})
    # Validate user has permissions to list reports
    # Call list_custom_reports with pagination parameters and user context
    # Return SavedReportListResponse with paginated reports
    return SavedReportListResponse(items=[], total=0, page=1, page_size=20, pages=0)


@router.post('/reports/execute', response_model=CustomReportResponse)
async def execute_report(
    request: CustomReportRequest,
    current_user: User = Depends(get_current_user)
):
    """Executes a custom report and returns the results"""
    logger.info("Received execute report request", extra={"additional_data": {"request": request}})
    # Validate user has permissions to execute the report
    # Call execute_custom_report with request parameters and user context
    # Return CustomReportResponse with the report results
    return CustomReportResponse(report_id=uuid.uuid4(), report_name="My Report", parameters={}, results={}, visualizations=[], execution_time=datetime.datetime.now())


@router.post('/reports/export')
async def export_report(
    request: CustomReportRequest,
    current_user: User = Depends(get_current_user)
):
    """Exports a custom report in the specified format"""
    logger.info("Received export report request", extra={"additional_data": {"request": request}})
    # Validate user has permissions to export the report
    # Call export_custom_report with request parameters and user context
    # Return export metadata including download URL
    return {"download_url": "https://example.com/report.xlsx"}


@router.get('/reports/fields')
async def get_report_fields(
    report_type: str,
    current_user: User = Depends(get_current_user)
):
    """Returns available fields for custom reports"""
    logger.info("Received get report fields request", extra={"additional_data": {"report_type": report_type}})
    # Validate user has permissions to access report fields
    # Call get_available_report_fields with report_type and user context
    # Return list of available fields with metadata
    return []